Fixed supervisor startup

hheimbuerger opened the pull request:
A simple change to get the supervisor working again.

Note that I don't know what the supervisor does or what it is supervising. So I haven't really tested anything, it just doesn't fail to start anymore:

```
neohabitat-neohabitat-1         | > pushserver@0.1.0 debug /neohabitat/pushserver
neohabitat-neohabitat-1         | > supervisor --inspect -i public,views -w bin,constants,habiproxy,routes -- bin/pushserver
neohabitat-neohabitat-1         |
neohabitat-neohabitat-1         | Missing/invalid defaults.elko configuration file. Proceeding with factory defaults.
neohabitat-neohabitat-1         | info: Habitat to Elko Bridge listening on 0.0.0.0:1337
neohabitat-neohabitat-1         |
neohabitat-neohabitat-1         | Running node-supervisor with
neohabitat-neohabitat-1         |   program '--inspect bin/pushserver'
neohabitat-neohabitat-1         |   --watch 'bin,constants,habiproxy,routes'
neohabitat-neohabitat-1         |   --ignore 'public,views'
neohabitat-neohabitat-1         |   --extensions 'node,js'
neohabitat-neohabitat-1         |   --exec 'node'
neohabitat-neohabitat-1         |
neohabitat-neohabitat-1         | Starting child process with 'node --inspect bin/pushserver'
neohabitat-neohabitat-1         | Ignoring directory '/neohabitat/pushserver/public'.
neohabitat-neohabitat-1         | Ignoring directory '/neohabitat/pushserver/views'.
neohabitat-neohabitat-1         | Watching directory '/neohabitat/pushserver/bin' for changes.
neohabitat-neohabitat-1         | Press rs for restarting the process.
neohabitat-neohabitat-1         | Watching directory '/neohabitat/pushserver/constants' for changes.
neohabitat-neohabitat-1         | Press rs for restarting the process.
neohabitat-neohabitat-1         | Watching directory '/neohabitat/pushserver/habiproxy' for changes.
neohabitat-neohabitat-1         | Press rs for restarting the process.
neohabitat-neohabitat-1         | Watching directory '/neohabitat/pushserver/routes' for changes.
neohabitat-neohabitat-1         | Press rs for restarting the process.
neohabitat-neohabitat-1         | Debugger listening on ws://127.0.0.1:9229/bc9c9b54-134b-4f9c-b859-49973341d393
neohabitat-neohabitat-1         | For help, see: https://nodejs.org/en/docs/inspector
neohabitat-neohabitat-1         | WebSocket settings:
neohabitat-neohabitat-1         |     - proxying from 0.0.0.0:1987 to qlink:1986
neohabitat-neohabitat-1         |     - Running in unencrypted HTTP (ws://) mode
```

StuBlad commented:
@hheimbuerger asked me to share the log of these changes not working when applied to @AmandaJonesAway branch.

`neohabitat_1         | Running node-supervisor with
neohabitat_1         |   program '/neohabitat/bridge/Habitat2ElkoBridge.js -l 0.0.0.0:1337 -e neohabitat:2018 -m neohabitatmongo:27017/elko -c context-Downtown_5f -t debug'
neohabitat_1         |   --watch 'bridge'
neohabitat_1         |   --extensions 'node,js,0:1337'
neohabitat_1         |   --exec 'node'
neohabitat_1         |
neohabitat_1         | Starting child process with 'node /neohabitat/bridge/Habitat2ElkoBridge.js -l 0.0.0.0:1337 -e neohabitat:2018 -m neohabitatmongo:27017/elko -c context-Downtown_5f -t debug'
neohabitat_1         | Watching directory '/neohabitat/bridge' for changes.
neohabitat_1         | Press rs for restarting the process.
neohabitat_1         |
neohabitat_1         | > pushserver@0.1.0 debug /neohabitat/pushserver
neohabitat_1         | > supervisor --inspect -i public,views -w bin,constants,habiproxy,routes -- bin/pushserver
neohabitat_1         |
neohabitat_1         | Missing/invalid defaults.elko configuration file. Proceeding with factory defaults.
neohabitat_1         |
neohabitat_1         | Running node-supervisor with
neohabitat_1         |   program '--inspect bin/pushserver'
neohabitat_1         |   --watch 'bin,constants,habiproxy,routes'
neohabitat_1         |   --ignore 'public,views'
neohabitat_1         |   --extensions 'node,js'
neohabitat_1         |   --exec 'node'
neohabitat_1         |
neohabitat_1         | Starting child process with 'node --inspect bin/pushserver'
neohabitat_1         | Ignoring directory '/neohabitat/pushserver/public'.
neohabitat_1         | Ignoring directory '/neohabitat/pushserver/views'.
neohabitat_1         | Watching directory '/neohabitat/pushserver/bin' for changes.
neohabitat_1         | Press rs for restarting the process.
neohabitat_1         | Watching directory '/neohabitat/pushserver/constants' for changes.
neohabitat_1         | Press rs for restarting the process.
neohabitat_1         | Watching directory '/neohabitat/pushserver/habiproxy' for changes.
neohabitat_1         | Press rs for restarting the process.
neohabitat_1         | Watching directory '/neohabitat/pushserver/routes' for changes.
neohabitat_1         | Press rs for restarting the process.
neohabitat_1         | info: Habitat to Elko Bridge listening on 0.0.0.0:1337
neohabitat_1         | Debugger listening on ws://127.0.0.1:9229/54851a49-172a-46c1-b5d4-1069201daacf
neohabitat_1         | For help, see: https://nodejs.org/en/docs/inspector
neohabitat_1         | internal/modules/cjs/loader.js:638
neohabitat_1         |     throw err;
neohabitat_1         |     ^
neohabitat_1         |
neohabitat_1         | Error: Cannot find module 'serve-favicon'
neohabitat_1         |     at Function.Module._resolveFilename (internal/modules/cjs/loader.js:636:15)
neohabitat_1         |     at Function.Module._load (internal/modules/cjs/loader.js:562:25)
neohabitat_1         |     at Module.require (internal/modules/cjs/loader.js:692:17)
neohabitat_1         |     at require (internal/modules/cjs/helpers.js:25:18)
neohabitat_1         |     at Object.<anonymous> (/neohabitat/pushserver/app.js:3:15)
neohabitat_1         |     at Module._compile (internal/modules/cjs/loader.js:778:30)
neohabitat_1         |     at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
neohabitat_1         |     at Module.load (internal/modules/cjs/loader.js:653:32)
neohabitat_1         |     at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
neohabitat_1         |     at Function.Module._load (internal/modules/cjs/loader.js:585:3)
neohabitat_1         | Program node --inspect bin/pushserver exited with code 1
neohabitat_1         |
neohabitat_1         | Starting child process with 'node --inspect bin/pushserver'
neohabitat_1         | Debugger listening on ws://127.0.0.1:9229/f160a104-f0e7-472a-98ef-997662bda53f
neohabitat_1         | For help, see: https://nodejs.org/en/docs/inspector
neohabitat_1         | internal/modules/cjs/loader.js:638
neohabitat_1         |     throw err;
neohabitat_1         |     ^`

hheimbuerger commented:
@StuBlad Thanks for sharing.

It looks like you simply haven't installed the `serve-favicon` package. Given that it's an essential part of the Dockerfile, I suspect you built your container from a different version of the Dockerfile.

I suggest you just rebuild the container once more (`docker compose build`). Observe the output of step 4 (+/-1). Does it show `npm install -g supervisor serve-favicon` as the command? If not, then you have a different Dockerfile than on Amanda's branch. If it does, then see whether there was an error message preventing the installation of `serve-favicon`.

StuBlad commented:
@hheimbuerger I cloned the most recent version of Amanda's fork and that's the Dockerfile that I'm using which contains the instruction to install the `serve-favicon` package.

`=> [ 4/12] RUN npm install -g supervisor serve-favicon          2.8s` is the output from running the build command, yet it still doesn't work and doesn't offer any error messages preventing the installation of the module.

[dockercomposebuild.txt](https://github.com/frandallfarmer/neohabitat/files/11153649/dockercomposebuild.txt)
[dockercomposeup.txt](https://github.com/frandallfarmer/neohabitat/files/11153650/dockercomposeup.txt)

Here is the full output of both commands.

You can see that when trying to bring things up, it does appear to be trying to install the modules and it's able to open the bridge:

```
neohabitat-neohabitat-1         | npm WARN neohabitat@0.1.0 No repository field.
neohabitat-neohabitat-1         |
neohabitat-neohabitat-1         | added 135 packages from 83 contributors and audited 135 packages in 9.24s
neohabitat-neohabitat-1         |
neohabitat-neohabitat-1         | 10 packages are looking for funding
neohabitat-neohabitat-1         |   run `npm fund` for details
neohabitat-neohabitat-1         |
neohabitat-neohabitat-1         | found 1 moderate severity vulnerability
neohabitat-neohabitat-1         |   run `npm audit fix` to fix them, or `npm audit` for details
neohabitat-neohabitat-1         | false
neohabitat-neohabitat-1         |  - Skipping cron..
neohabitat-neohabitat-1         |  - Enabling JMX debugging for Neohabitat...
neohabitat-neohabitat-1         |  - Running Neohabitat Elko server with args: org.elkoserver.server.context.ContextServerBoot trace_cont=DEBUG trace_comm=EVENT tracelog_dir=/plz/log/to/stdout tracelog_tag=context conf.listen.host=neohabitat:9000 conf.listen.bind=neohabitat:9000 conf.listen.protocol=tcp conf.comm.jsonstrictness=true conf.context.entrytimeout=999999 conf.context.odb=mongo conf.context.odb.mongo.hostport=neohabitatmongo:27017 conf.context.objstore=org.elkoserver.objdb.store.mongostore.MongoObjectStore conf.context.name=neohabitat conf.context.shutdownpassword=figleaf conf.msgdiagnostics=true
neohabitat-neohabitat-1         |
neohabitat-neohabitat-1         | Running node-supervisor with
neohabitat-neohabitat-1         |   program '/neohabitat/bridge/Habitat2ElkoBridge.js -l 0.0.0.0:1337 -e neohabitat:2018 -m neohabitatmongo:27017/elko -c context-Downtown_5f -t debug'
neohabitat-neohabitat-1         |   --watch 'bridge'
neohabitat-neohabitat-1         |   --extensions 'node,js,0:1337'
neohabitat-neohabitat-1         |   --exec 'node'
neohabitat-neohabitat-1         |
neohabitat-neohabitat-1         | Starting child process with 'node /neohabitat/bridge/Habitat2ElkoBridge.js -l 0.0.0.0:1337 -e neohabitat:2018 -m neohabitatmongo:27017/elko -c context-Downtown_5f -t debug'
neohabitat-neohabitat-1         | Watching directory '/neohabitat/bridge' for changes.
neohabitat-neohabitat-1         | Press rs for restarting the process.```

hheimbuerger commented:
@StuBlad I'm not quite following. You posted two different startup logs from the supervisor, once as a link (`dockercomposeup.txt`) and once as a code block. What's the difference between the two?

It looks like one had the `serve-favicon` package installed and one didn't. But I don't understand what you did differently and why you're posting the one where it wasn't installed?

StuBlad commented:
Sorry, I posted the same log but I shared the code block as an excerpt from `dockercomposeup.txt` incase you just wanted to see what I believed to be the relevant part of the log.

They both don't have the `serve-favicon` package installed properly. The except in the code block happens just before the server is about to start and you can see npm installed packages, however immediately after, it tries to launch the pushserver which then gives the `serve-favicon` error constantly.

It feels like somehow, the NeoHabitat server and the pushserver are using different installations of npm possibly, or at least they are both looking in different places. The global flag being specified for the NeoHabitat server may then not be affecting whatever the pushserver is looking at.

I'm not sure how to tell where both of those services are looking for their packages from or how to rectify that issue.

hheimbuerger commented:
> They both don't have the `serve-favicon` package installed properly. The excerpt in the code block happens just before the server is about to start and you can see npm installed packages, however immediately after, it tries to launch the pushserver which then gives the `serve-favicon` error constantly.

That sounds like a reasonable interpretation.

Not sure why I didn't experience this, though. I cannot test right now, but distinctly remember that I fixed this as a very isolated issue, i.e. I'm almost certain I added that one line to the Dockerfile, rebuilt, and the error went away.
Maybe I'm misremembering, because your explanation sounds more likely.

frandallfarmer commented:
Thank you for updating these problems.
